modules/apigee_edge_apidocs/src/ApiDocSpecFetcher.php

```diff
@@ -0,0 +1,262 @@
+      if (empty($data)) {
+        $this->log($apidoc, static::TYPE_ERROR, 'API Doc %label: OpenAPI specification file located at %url is empty.', [
+          '%url' => $file_uri,
+          '%label' => $apidoc->label(),
+        ], $show_messages);
+        return FALSE;
+      }
+
+      // Only save file if it hasn't been fetched previously.
+      $data_md5 = md5($data);
+      $prev_md5 = $apidoc->get('spec_md5')->isEmpty() ? NULL : $apidoc->get('spec_md5')->value;
+      if ($prev_md5 != $data_md5) {
+        $filename = $this->fileSystem->basename($file_uri);
+        $specs_definition = $apidoc->getFieldDefinition('spec')->getItemDefinition();
+        $target_dir = $specs_definition->getSetting('file_directory');
+        $uri_scheme = $specs_definition->getSetting('uri_scheme');
+        $destination = "$uri_scheme://$target_dir/";
+
+        try {
+          $this->checkRequirements($destination);
+          $file = file_save_data($data, $destination . $filename, FILE_EXISTS_RENAME);
+
+          if (empty($file)) {
+            throw new \Exception('Could not save API Doc specification file.');
+          }
+        }
+        catch (\Exception $e) {
+          $this->log($apidoc, static::TYPE_ERROR, 'Error while saving API Doc spec file from URL on API Doc ID: %id. Error: %error', [
+            '%id' => $apidoc->id(),
+            '%error' => $e->getMessage(),
+          ], $show_messages);
+          return FALSE;
+        }
+
+        $spec_value = [
+            'target_id' => $file->id(),
+          ] + $spec_value;
+        $apidoc->set('spec', $spec_value);
+        $apidoc->set('spec_md5', $data_md5);
+        $apidoc->set('fetched_timestamp', time());
+
+        $needs_save = TRUE;
+      }
+    }
+
+    elseif ($apidoc->get('spec_file_source')->value == ApiDocInterface::SPEC_AS_FILE) {
+      $this->log($apidoc, static::TYPE_STATUS, 'API Doc %label is using a file upload as source. Nothing to update.', [
+        '%label' => $apidoc->label(),
+      ], $show_messages);
+      $needs_save = FALSE;
+    }
+
+    return $needs_save;
+  }
+
+  /**
+   * Log a message, and optionally display it on the UI.
+   *
+   * @param \Drupal\apigee_edge_apidocs\Entity\ApiDocInterface $apidoc
+   *   The API Doc entity.
+   * @param string $type
+   *   Type of message.
+   * @param string $message
+   *   The message.
+   * @param array $params
+   *   Optional parameters array.
+   * @param bool $show_messages
+   *   TRUE if message should be displayed to the UI as well.
+   */
+  private function log(ApiDocInterface $apidoc, string $type, string $message, array $params = [], bool $show_messages = TRUE) {
+    switch ($type) {
+      case static::TYPE_ERROR:
+        $this->logger->error($message, $params);
+        break;
+      case static::TYPE_STATUS:
+      default:
+        $this->logger->info($message, $params);
+    }
+    if ($show_messages) {
+      $this->messenger->addMessage($this->t($message, $params), $type);
```

Comment:
I do not think this solves the problem with internationalization.
You can easily verify whether this work or not enable internationalization and try to translate `API Doc %label is using a file upload as source. Nothing to update.`.